components/compositing/compositor.rs

```diff
@@ -475,55 +473,56 @@ impl<Window: WindowMethods> IOCompositor<Window> {
                 self.scroll_fragment_to_point(scroll_root_id, point);
             }
 
-            (Msg::MoveTo(point),
+            (Msg::MoveTo(top_level_browsing_context_id, point),
              ShutdownState::NotShuttingDown) => {
-                self.window.set_position(point);
+                self.window.set_position(top_level_browsing_context_id, point);
             }
 
-            (Msg::ResizeTo(size),
+            (Msg::ResizeTo(top_level_browsing_context_id, size),
              ShutdownState::NotShuttingDown) => {
-                self.window.set_inner_size(size);
+                self.window.set_inner_size(top_level_browsing_context_id, size);
             }
 
-            (Msg::GetClientWindow(send),
+            (Msg::GetClientWindow(top_level_browsing_context_id, send),
              ShutdownState::NotShuttingDown) => {
-                let rect = self.window.client_window();
+                let rect = self.window.client_window(top_level_browsing_context_id);
                 if let Err(e) = send.send(rect) {
                     warn!("Sending response to get client window failed ({}).", e);
                 }
             }
 
-            (Msg::Status(message), ShutdownState::NotShuttingDown) => {
-                self.window.status(message);
+            (Msg::Status(top_level_browsing_context_id, message),
+             ShutdownState::NotShuttingDown) => {
+                self.window.status(top_level_browsing_context_id, message);
             }
 
-            (Msg::LoadStart, ShutdownState::NotShuttingDown) => {
-                self.window.load_start();
+            (Msg::LoadStart(top_level_browsing_context_id), ShutdownState::NotShuttingDown) => {
+                self.window.load_start(top_level_browsing_context_id);
```

Comment:
irc discussion: http://logs.glob.uno/?c=mozilla%23servo#c727681

ok - let me try to describe the problem you're describing, to make sure we are on the same page:

- Load A happens, LoadStart is fired
- Load B happens right after, LoadStart is fired
- A finishes to load
- B finishes to load

We don't want to be in a situation where we fire LoadComplete when A finished loading, because in that case, the embedder would not expect another LoadComplete and would assume it's B that finished loading.

My assumption is that we can only end up in these situations:

```
> Load A
> Load B
< LoadStart A
< LoadComplete A
< LoadStart B
< LoadComplete B
```

or

```
> Load A
> Load B
< LoadStart A
< LoadStart B
< LoadComplete B
```

What we don't want is

```
> Load A
> Load B
< LoadStart A
< LoadStart B
< LoadComplete A
< LoadComplete B
```

The intended logic of the code is to only fire LoadComplete for X if X is the current pipeline.

So first, this is wrong, because we can be in a situation where A is still the current pipeline because B is still pending. So my PR is wrong here. We also need to check the pending changes and make sure nothing else is expected.

Next, can message latency be an issue? Which is I think what you were getting at. Is it possible to get LoadComplete for A inside the constellation and not know that embedder requested to load B? Yes. But, what really matters is to not fire LoadComplete A after we fired **LoadStart** B. The embedder relies on the LoadStart message, not on when it calls Load A / Load B.

And LoadStart and LoadComplete are states known by the constellation. There is not reason for this to have any race here.

The only real issue would be if we get messages unordered, which I hope is not the case :)

Conclusion: I need to check for pending changes before firing LoadComplete. And I believe the asynchronous nature of the messages is not a problem as long as the embedder relies on LoadStart.

I hope this clarify the situation.

Let me know.